emission/analysis/classification/inference/labels/pipeline.py

```diff
@@ -55,48 +57,67 @@ def run_prediction_pipeline(self, user_id, time_range):
+        cleaned_trip_list = self.toPredictTrips
+        inferred_trip_list = []
+        results_dict = {}
+        ensemble_list = []
+
+        # Create list of inferred trips
+        for cleaned_trip in cleaned_trip_list:
             cleaned_trip_dict = copy.copy(cleaned_trip)["data"]
             inferred_trip = ecwe.Entry.create_entry(user_id, "analysis/inferred_trip", cleaned_trip_dict)
-            
-            # Run the algorithms and the ensemble, store results
-            results = self.compute_and_save_algorithms(inferred_trip)
-            ensemble = self.compute_and_save_ensemble(inferred_trip, results)
+            inferred_trip_list.append(inferred_trip)
+    
+        # Computing outside loop by passing trip_list to ensure model loads once
+        # Run the algorithms and the ensemble, store results
+        results_dict = self.compute_and_save_algorithms(user_id, inferred_trip_list)
+        ensemble_list = self.compute_and_save_ensemble(inferred_trip_list, results_dict)
 
+        start_insert_inferred_trip_time = time.process_time()
+        for cleaned_trip, inferred_trip, ensemble in zip(cleaned_trip_list, inferred_trip_list, ensemble_list):
             # Put final results into the inferred trip and store it
             inferred_trip["data"]["cleaned_trip"] = cleaned_trip.get_id()
             inferred_trip["data"]["inferred_labels"] = ensemble["prediction"]
             self.ts.insert(inferred_trip)
 
             if self._last_trip_done is None or self._last_trip_done["data"]["end_ts"] < cleaned_trip["data"]["end_ts"]:
                 self._last_trip_done = cleaned_trip
+        print(f"{arrow.now()} Inside run_prediction_pipeline: Saving inferred_trip total time = {time.process_time() - start_insert_inferred_trip_time}")
     
     # This is where the labels for a given trip are actually predicted.
     # Though the only information passed in is the trip object, the trip object can provide the
     # user_id and other potentially useful information.
-    def compute_and_save_algorithms(self, trip):
-        predictions = []
+    def compute_and_save_algorithms(self, user_id, trip_list):
+        predictions_dict = {trip.get_id(): [] for trip in trip_list}
         for algorithm_id, algorithm_fn in primary_algorithms.items():
-            prediction = algorithm_fn(trip)
-            lp = ecwl.Labelprediction()
-            lp.trip_id = trip.get_id()
-            lp.algorithm_id = algorithm_id
-            lp.prediction = prediction
-            lp.start_ts = trip["data"]["start_ts"]
-            lp.end_ts = trip["data"]["end_ts"]
-            self.ts.insert_data(self.user_id, "inference/labels", lp)
-            predictions.append(lp)
-        return predictions
+            prediction_list = algorithm_fn(user_id, trip_list)
+            start_insert_inference_labels_time = time.process_time()
+            for trip, prediction in zip(trip_list, prediction_list):
+                lp = ecwl.Labelprediction()
```

Comment:
**future fix??**: my one comment here is that if we standardize on an interface for the `algorithm_fn`, I believe we can simplify this further.

This can become (pseudocode):

```
for algorithm_id, algorithm_fn in primary_algorithms.items():
   algo_model = algorithm.load_model()
   for trip in trip_list:
      prediction = model.predict(trip)
      # and then the original code
```

Back when this code was written, `trip_model`, with its clean interfaces to load, predict and save, had not yet been implemented. So `algorithm_fn` was supposed to be an abstraction layer of the multiple algorithms.

But now that we have the abstractions in trip_model, I am not sure that we need two levels of abstraction. I am open to deferring the cleanup until the next round, but I am also open to just removing one level of abstraction right now, since that will result in far fewer changes.

Reply:
Right now, there is a map that has algorithm name to algorithm fn and then the algorithm fn (in our case) calls `predict_with_n`. instead, the map can have algorithm name to dict of model parameters, and then inside the for loop, we can just load the model and continue to predict trip by trip and not load the model within the predict function (which works, but is not pretty).

Alternatively, we can keep the algorithm function, `eacili.predict_cluster_confidence_discounting` but load the model in there. But I don't see the need for the second level of abstraction if we assume that all our algorithms will follow the L -> P -> S workflow.

Reply:
If you are not comfortable with removing the `algorithm_fn` abstraction, you can simply do the following 2 lines of code to be changed.

```
def predict_cluster_confidence_discounting(trip, max_confidence=None, first_confidence=None, confidence_multiplier=None):
    # load application config
    model_type = eamtc.get_model_type()
    model_storage = eamtc.get_model_storage()
    ### load model  <------ NEW CODE
    labels, n = eamur.predict_labels_with_n(trip, model) <----- CHANGED CODE
    if n <= 0:  # No model data or trip didn't match a cluster
        logging.debug(f"In predict_cluster_confidence_discounting: n={n}; returning as-is")
        return labels

    confidence_coeff = n_to_confidence_coeff(n, max_confidence, first_confidence, confidence_multiplier)
    logging.debug(f"In predict_cluster_confidence_discounting: n={n}; discounting with coefficient {confidence_coeff}")

    labels = copy.deepcopy(labels)
    for l in labels: l["p"] *= confidence_coeff
    return labels
```

Reply:
I have a few queries which are still confusing to me.

> But now that we have the abstractions in trip_model, I am not sure that we need two levels of abstraction.

1.  By "two levels", I believe you are referring to: 
```
eacili.predict_cluster_confidence_discounting -> eamur.predict_labels_with_n -> model.predict()
```
I see that trip_model is an abstract class and one child class / concrete implementation is GreedySimilarityBinning class. 
**Do we intend on removing the calls to the first two functions in the above chain and directly call only model.predict() which is in effect a call to GreedySimilarityBinning.predict() depending on the model parameters?**

------------

> Right now, there is a map that has algorithm name to algorithm fn and then the algorithm fn (in our case) calls `predict_with_n`. instead, the map can have algorithm name to dict of model parameters...

> Alternatively, we can keep the algorithm function, `eacili.predict_cluster_confidence_discounting` but load the model in there. But I don't see the need for the second level of abstraction if we assume that all our algorithms will follow the L -> P -> S workflow.

2. I believe I've understood that the same trip model will be loaded for each user irrespective of algorithm if we move up the model loading, model parameters up into `eacilp.pipeline`.

If this is correct, then yes, using a map of algorithm_name to dict_model_parameters is useful:
```
primary_algorithms = {
    ecwl.AlgorithmTypes.CONFIDENCE_DISCOUNTED_CLUSTER: dict_model_params
}

# In the for loop
   # Load model
   for trip in trip_list:
      prediction = model.predict(trip)
```

But in doing this (in continuation with my 1st question), this will call model.predict(), and I fail to see how our algorithm_name will have a unique implementation if we are removing the call to its respective function?
For instance, on removing the call to `eacili.predict_cluster_confidence_discounting`, we are no longer doing the `confidence_coeff`calculations. 
**So what would be the significance of the algorithm name in the map in this case?**

We also have placeholder_predictor() functions in inferrers.py which are being called in TestLabelInferencePipeline.py through the compute_and_save_algorithms() as well.
I believe for these functions, loading model and making predictions won't be useful as these don't load the model and just return predictions.